src/compiler/scala/tools/nsc/typechecker/Implicits.scala

```diff
@@ -221,14 +221,22 @@ trait Implicits {
     }
 
     def isCyclicOrErroneous: Boolean = {
-      if (!isCyclicOrErroneousCache.isKnown) isCyclicOrErroneousCache = computeIsCyclicOrErroneous
-      isCyclicOrErroneousCache.booleanValue
+      if (!isCyclicOrErroneousCache.isKnown)
+        try global.withPropagateCyclicReferences {
+          val cyclicOrErroneous = sym.hasFlag(LOCKED) || containsError(tpe)
+          isCyclicOrErroneousCache = cyclicOrErroneous
+          cyclicOrErroneous
+        } catch {
+          case _: CyclicReference =>
```

Comment:
I believe that this is a correct test. If it's unavoidably a cyclic reference, then `sym` should be locked... otherwise we probably should consider it next time around.

I'm going to compile shapeless w/ and w/o this patch to see if this will cause a major performance regression. I believe it won't, however.